map.js

```diff
@@ -467,66 +415,108 @@ function initListeners(){
+    const ICON_NATIVE_HEIGHT = 20;
+    let oldZoom = zoomLevel;
+    zoomLevel += deltaZ;
+    //clamp zoom
+    if(zoomLevel > maxZoom) zoomLevel = maxZoom;
+    if(zoomLevel < minZoom) zoomLevel = minZoom;
+
+    if(oldZoom == zoomLevel){
+        return;
+    }
+
+    //adjust icon size
+    var m_iconH = ICON_NATIVE_HEIGHT / zoomLevel;
+    if(zoomLevel > 1.75)m_iconH = ICON_NATIVE_HEIGHT / zoomLevel * 2;
+    else if(zoomLevel > 1.5)m_iconH = ICON_NATIVE_HEIGHT / zoomLevel * 1.5;
+    else if(zoomLevel > 1.25)m_iconH = ICON_NATIVE_HEIGHT / zoomLevel * 1.25;
+    _iconH = m_iconH;
+
+    //make map zoom in to zoomPoint.
+    //1: calculate current zoomPoint in map coords
+    //2. calculate where that point is on the new map
+    //3. calculate where the corner needs to be to set that point as the center
+    let oldCenterMapCoord = screenSpaceToMapSpace(zoomPoint);
+    let newCenterMapCoord = oldCenterMapCoord.multiply(oldZoom/zoomLevel);
+    let newCornerMapCoord = newCenterMapCoord.subtract(zoomPoint);
+
+    //moveMap takes a delta, so we subtract new from old. 
+    moveMap(screenOriginInMapCoords.subtract(newCornerMapCoord));
+}
+
 //converts worldspace cords into map coords.
 //this is a pixel measurement from upper left of map image.
-function worldSpaceToMapSpace(x = 0, y = 0){
+function worldSpaceToMapSpace(point){
     //first, we convert world space into map space.
     var MapW = img_Map.width;
     var MapH = img_Map.height;
     const worldW = 480000;
     const worldH = 400000;
-    const ICON_NATIVE_HEIGHT = 20;
     
     //world coords are -240,000 to 240,000 in the x direction
     //and -200,000 to 200,000 in the y direction
 
     //for most things, we store the "map coords", and then that gets converted to viewport(aka canvas) coords with simple vector addition at draw time.
 
     //first, convert to positive number between 0 and 1.
-    let fraction_x = (Math.round(x) + worldW / 2) / worldW;
-    let fraction_y = (-Math.round(y) + worldH / 2) / worldH;
+    let fraction_x = (Math.round(point.x) + worldW / 2) / worldW;
+    let fraction_y = (-Math.round(point.y) + worldH / 2) / worldH;
 
-    //then, adjust by zoom.
-    var m_iconH = ICON_NATIVE_HEIGHT / zoomLevel;
-    if(zoomLevel > 1.75)m_iconH = ICON_NATIVE_HEIGHT / zoomLevel * 2;
-    else if(zoomLevel > 1.5)m_iconH = ICON_NATIVE_HEIGHT / zoomLevel * 1.5;
-    else if(zoomLevel > 1.25)m_iconH = ICON_NATIVE_HEIGHT / zoomLevel * 1.25;
-
-    //now set the map space x and y to the scaled coords.
+    //then adjust for the new map height/width.
     let map_x = (MapW * fraction_x) / zoomLevel;
     let map_y = (MapH * fraction_y) / zoomLevel;
-    iconH = m_iconH;
-    return {x:map_x, y:map_y, iconH:m_iconH}
+
+    return new Point(map_x, map_y);
+}
+
+/**
+ * Convert a point in map space to a point in screen space.
+ * @param {Point} mapSpacePoint 
+ * @returns {Point} screen space point
+ */
+function mapSpaceToScreenSpace(mapSpacePoint){
```

Comment:
Special functions to convert from coordinate space A to B and B to A because its super easy to get confused which coordinate space you're in, and how to get to a different one.